### ISSUE TITLE
HTTP/3 references QUIC Stream IDs directly, allowing illegal references

### ISSUE THREAD

**Issue by ianswett** [quicwg/base-drafts#3273]
Currently, HTTP/3 references QUIC Stream IDs directly in the case of requests.  However, requests are only 1 of the 4 stream types, so 3/4 of the values in GOAWAY are currently invalid, though we don't specify what happens if you receive one of those invalid values in the current draft(or I missed it).

QUIC transport has moved to a design for MAX_STREAMS(https://tools.ietf.org/html/draft-ietf-quic-transport-24#section-19.11) that specifies the allowed number of streams of a given type, rather than the Stream ID.

Currently, this is a very minor point, since it's only the GOAWAY frame that references Stream ID, but future frames, such as the PRIORITY_UPDATE frame from the HTTP priorities draft will also want to reference the Stream ID, as will future extension frames I expect.

So I suggest we call this a Request ID and it's the Stream ID of a client bidi stream right shifted 2 bits or however else we'd like to describe it.

**Comment by kazuho** [comment]
While I agree that such a design is possible, I am in favor of using stream IDs as they are, because that is what we have done in HTTP/2.

To be precise, HTTP/2 has stream 0 (which is used for transmitting control messages), odd-numbered streams (for carrying requests), even-numbered streams (for carrying pushed requests).

To paraphrase, we have had these ID holes in HTTP/2, and we haven't had issues (I assume). We've already built our infrastructure around this design, for example, we simply log the stream ID and hope that that would be sufficient to identify the request, regardless of it being one initiated by the client or one being pushed by the server.

Introducing the concept of "request ID" would be a departure from this design in HTTP/2. As there would be a different number space for pushed requests, we'd be required to change the logging structure to convey a tuple of request-type and request-ID. While that is not impossible, I am not sure if that change is worth the cost.

**Comment by RyanAtGoogle** [comment]
I think this issue came up in Cupertino and the current approach was about as close to consensus as any of the alternatives. I don't feel strongly either way which makes me somewhat inclined to leave things alone.

**Comment by dtikhonov** [comment]
> Introducing the concept of "request ID" would be a departure from this design in HTTP/2. As there would be a different number space for pushed requests, we'd be required to change the logging structure to convey a tuple of request-type and request-ID. While that is not impossible, I am not sure if that change is worth the cost.

Not to mention having to maintain a Request ID -> Stream ID mapping.  I agree with @kazuho.

**Comment by martinthomson** [comment]
This suggestion was considered and discussed already as part of the decision to add push identifiers.  I don't see any new information here.

**Comment by ianswett** [comment]
This seems a bit at odd with our design principles, since it requires the HTTP/3 implementation to understand the QUIC Stream ID numbering scheme.   In our implementation that's viable and it probably is in others.  The current text says:
  "A client MUST treat receipt of a GOAWAY
   frame containing a Stream ID of any other type as a connection error
   of type H3_ID_ERROR."

I don't remember us discussing this particular detail of GOAWAY in Cupertino, but I also wasn't very invested in the GOAWAY discussion, so I believe I missed it.

To be clear, the Request ID -> Stream ID mapping is multiplying by 4 in this case.  There is no state required.

TLDR: This seems weird, but I'm ok with closing with no action if others don't care to fix the oddity.

**Comment by RyanAtGoogle** [comment]
> I don't remember us discussing this particular detail of GOAWAY in Cupertino, but I also wasn't very invested in the GOAWAY discussion, so I believe I missed it.

I think it was on the last day:

https://github.com/quicwg/wg-materials/blob/master/interim-19-10/minutes.md#next-issue-goaway-see-above

**Comment by LPardue** [comment]
I find it easier to refer to Stream IDs because that's what STOP_SENDING and RESET_STREAM use. Having some other value sent on the wire seems like a good way to get interop problems, especially because the first request would have stream ID 0 and 0*4 = 0.

**Comment by larseggert** [comment]
Closed with no action, as indicated by @ianswett

**Comment by MikeBishop** [comment]
> This seems a bit at odd with our design principles, since it requires the HTTP/3 implementation to understand the QUIC Stream ID numbering scheme.

While the issue is closed, I wanted to point out that it does not imply this requirement.  I've held very strictly to this mental model:

- HTTP/3 can ask QUIC for a stream of a given type
- HTTP/3 can query the newly-received stream for its ID
- HTTP/3 can thereafter use the ID to refer to the stream
- HTTP/3 can query QUIC for the type of a given Stream ID and/or retrieve a stream by ID and then query its type

If QUIC were to change how Stream IDs are allocated, the changes could be entirely within the transport.  Now, it's a fair critique to say that this model is not explicitly stated in the draft, and so it's not guaranteed that any given implementation would actually preserve this flexibility.

**Comment by ianswett** [comment]
I was aware of the first 3, but It's the last requirement that I wasn't sure of:

* HTTP/3 can query QUIC for the type of a given Stream ID and/or retrieve a stream by ID and then query its type

In this case, in order to ensure the GOAWAY doesn't reference an invalid Stream ID, you need to be able to query the type of a Stream ID even if the stream hasn't been created.

**Comment by kazuho** [comment]
I think I agree with @ianswett.

One trivial example is the value of GOAWAY.stream_id when the server closes an HTTP/3 connection that has carried no request. In such case, it is my understanding that GOAWAY.stream_id should be zero. But how does the server determine that value?

**Comment by martinthomson** [comment]
There is I think an additional assumption here: that stream IDs are allocated in ascending order.  More or less.  That isn't a hard requirement, but it informs how effective GOAWAY is.

```
global varint max_request_stream_id = 0;

on handling(request):
    max_request_stream_id = max(max_request_stream_id, request.stream_id)

on sending(goaway):
    goaway.stream_id = max_request_stream_id
    for r in open requests:
        if r.stream_id < max_request_stream_id:
            r.reset()
```

See, not especially complicated, but the extent to which you have to reset streams depends on the order in which stream IDs are allocated.

**Comment by LPardue** [comment]
@ianswett 
> In this case, in order to ensure the GOAWAY doesn't reference an invalid Stream ID, you need to be able to query the type of a Stream ID even if the stream hasn't been created.

I don't see why this is a problem - the stream ID value indicates the type even if an instance of that stream does not exist in a connection. In fact, we rely on this when GOAWAY with MAX_CLIENT_INITIATED_BIDI is sent. 

@kazuho 
> One trivial example is the value of GOAWAY.stream_id when the server closes an HTTP/3 connection that has carried no request. In such case, it is my understanding that GOAWAY.stream_id should be zero. But how does the server determine that value?

Spec says: 

> This identifier MAY be zero if no requests were processed.

 I think MT's example is pretty straightforward way to account

**Comment by kazuho** [comment]
@martinthomson I'm not sure if the logic is correct. It is my understanding that the server need to set GOAWAY.stream_id to the largest request stream ID it has processed, __added by 4__.

This is because the rules are:
* The GOAWAY frame indicates that client-initiated requests on lower stream IDs were or might be processed in this connection, while requests on the indicated stream ID and greater were rejected ([section 5.2](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-connection-shutdown)).
* It carries a QUIC Stream ID for a client-initiated bidirectional stream encoded as a variable-length integer ([section 7.2.6](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-goaway)).

@LPardue My point is that the specification requires a server to send an stream ID it has never seen, when there are no requests inflight.

I thought that it would be good to use as an example the case where no requests have ever been sent, and that the spec has to call out what should happen in that particular case is a leak.

But that might not have been the best example. Maybe I should have given the example when one request is exchanged then the server idle-timeouts. In such case, GOAWAY.stream_id is expected to be 4, which is a stream ID that the server has never seen.

**Comment by martinthomson** [comment]
As long as we don't rely on the stream ID carrying its type (that is, we rely on GOAWAY implicitly identifying the type of stream) we can use +1 rather than +4.  Otherwise, yeah, that exposes the internals of the transport.

**Comment by kazuho** [comment]
> we can use +1 rather than +4

That might be the reasonable thing to do. Though I do not care much, because there would be other hidden assumptions regardless: e.g., request stream ID can be represented using a 62-bit unsigned number starting from zero, request stream ID monotonically increases as the client uses them.

**Comment by martinthomson** [comment]
There is an assumption that only stream IDs that fit in a 62-bit number work, sure.  The monotonic increase isn't a strong assumption, only that there is some advantage to dividing the space into two with GOAWAY is useful.  You can do that split with totally random allocation of stream IDs and get some advantage; it's only if the stream IDs are monotonically *decreasing* that we get zero utility from GOAWAY.
